File: image/build-bitcode.py
#!/usr/bin/python3

import hashlib
import os
import subprocess
import sys

# Commands we may use.
PRT = "prt-get"
GETBC = "get-bc"
DIS = "llvm-dis"
PKGINFO = "pkginfo"
LOC = "loc"
OPT = "opt"

BC_DIR = ""
try:
    BC_DIR = os.environ["BC_DIR"]
except:
    print("BC_DIR environment variable not set!")
    sys.exit(3)

# Bitcode info file.
BC_INFO = BC_DIR + "/info.txt"

try:
    os.mkdir(BC_DIR)
except FileExistsError:
    pass
except Exception as e:
    print("Could not make $BC_DIR: {}".format(e))

PKGS_FILE = ""
try:
    PKGS_FILE = os.environ["PKGS_FILE"]
except:
    print("PKGS_FILE environment variable not set!")
    sys.exit(3)

PKGMK_CONF = "/etc/pkgmk.conf"

# First four bytes of ELF files.
ELF_MAGIC = bytes.fromhex("7F454c46")

pkgs = []
deps = set()

# Maps MD5 hashes to sets of installed files with that hash
aliases = {}
# Maps MD5 hashes to a "canonical" filename.
canon_fn = {}
# Maps MD5 hashes to the LOC of that BC.
locs = {}

cflags = ""
cxxflags = ""

# Set CFLAGS and CXXFLAGS from pkgmk.conf. Some ports don't respect pkgmk.conf's.
try:
    with open(PKGMK_CONF, "r") as f:
        for l in f.readlines():
            l = l.strip()
            if len(l) >= 6 and l[:13] == "export CFLAGS":
                parts = l.split("=")
                if len(parts) > 1:
                    cflags = "".join(parts[1:])
                    os.environ["CFLAGS"] = cflags
                    print("Exported CFLAGS")
            if len(l) >= 8 and l[:15] == "export CXXFLAGS":
                parts = l.split("=")
                if len(parts) > 1:
                    cxxflags = "".join(parts[1:])
                    os.environ["CXXFLAGS"] = cxxflags
                    print("Exported CXXFLAGS")
except:
    print("Did not export CFLAGS/CXXFLAGS")
    pass

try:
    pkgs_f = open(PKGS_FILE, "r")
except Exception as e:
    print("Could not open packages file: {}".format(e))
    sys.exit(1)
else:
    with pkgs_f:
        for pkg in pkgs_f:
            pkg = pkg.strip()

            # Check the package exists.
            if subprocess.run([PRT, "info", pkg], stdout=subprocess.DEVNULL).returncode != 0:
                print("Bad package name: {}".format(pkg))
                sys.exit(2)

            # We want the bitcode of pkg.
            pkgs.append(pkg)

            pkg_deps = subprocess.run([PRT, "depends", pkg], capture_output=True)
            # Skip the first line because it is some output info for the user.
            for dep in pkg_deps.stdout.strip().split(b"\n")[1:]:
                deps.add(dep.split()[1])

for pkg in pkgs:
    # Build + install the package.

    # If it's already installed, we update, otherwise, depinst.
    command = "update"
    if subprocess.run([PRT, "isinst", pkg]).returncode != 0:
        command = "depinst"

    if subprocess.run([PRT, "-if", "-im", "-is", "-kw", "-fr", "--install-scripts", command, pkg]).returncode != 0:
        print("Failed to build {}!".format(pkg))
        sys.exit(4)

    # Grab bitcode from binaries. We're looking for ELF files.
    footprint = subprocess.run([PKGINFO, "-l", pkg], capture_output=True)
    for installed_file in footprint.stdout.strip().split(b"\n"):
        # These files are relative to the package; make them relative to root.
        installed_file = "/" + installed_file.decode("utf-8")

        # Don't need directories, obviously.
        if os.path.isdir(installed_file):
            continue

        with open(installed_file, "rb") as f:
            bc_basename_path = BC_DIR + "/" + os.path.basename(installed_file)
            bc_basename = os.path.basename(bc_basename_path) + ".bc"
            if f.read(4) == ELF_MAGIC:
                md5 = hashlib.md5(f.read()).hexdigest()
                if md5 not in canon_fn:
                    canon_fn[md5] = bc_basename
                    # This wouldn't exist yet if we are here.
                    aliases[md5] = set()
                else:
                    aliases[md5].add(bc_basename)
                    # Already processed the canon file; this is the same.
                    continue

                get = subprocess.run([GETBC, "-o", bc_basename_path + ".bc",
                                      installed_file],
                                     capture_output=True)
                if get.returncode != 0:
                    print("Failed to get BC for {}: {}".format(installed_file, get.stdout))
                    sys.exit(5)

                # Grab .ll (for loc).
                if subprocess.run([DIS, bc_basename_path + ".bc"]).returncode != 0:
                    print("Failed to get disassembled {}".format(bc_basename_path + ".bc"))
                    sys.exit(6)

                # We grabbed the ll with debug info, so we'll strip it.
                # Result is a file of the same name.
                os.rename(bc_basename_path + ".bc", bc_basename_path + ".dbg.bc")
                if subprocess.run([OPT, "-strip-debug",
                                   bc_basename_path + ".dbg.bc",
                                   "-o",
                                   bc_basename_path + ".bc"]).returncode != 0:
                    print("Failed to strip debug info for {}".format(bc_basename_path + ".bc"))
                    sys.exit(7)

                # Delete debug BC.
                os.remove(bc_basename_path + ".dbg.bc")

                # Run the loc script and delete the .ll.
                loc = subprocess.run([LOC, bc_basename_path + ".ll"], capture_output=True)
                os.remove(bc_basename_path + ".ll")
                loc = loc.stdout.strip().decode("ascii")
                locs[md5] = loc

try:
    bc_info = open(BC_INFO, "a")
except Exception as e:
    print("Could not open bitcode info file: {}".format(e))
    sys.exit(8)

if cflags == "":
    cflags = "No user-defined CFLAGS"
if cxxflags == "":
    cxxflags = "No user-defined CXXFLAGS"

with bc_info:
    bc_info.write("CFLAGS\n")
    bc_info.write("------\n")
    bc_info.write(cflags + "\n\n")

    bc_info.write("CXXFLAGS\n")
    bc_info.write("--------\n")
    bc_info.write(cxxflags + "\n\n")

    bc_info.write("LOC\t\tBitcode\t\tAliases\n")
    bc_info.write("---\t\t-------\t\t-------\n")

    for md5 in canon_fn.keys():
        fn = canon_fn[md5]
        loc = locs[md5]
        fn_aliases = aliases[md5]

        bc_info.write(loc + "\t\t" + fn + "\t\t")
        first = True
        for alias in fn_aliases:
            if not first:
                bc_info.write(" ")

            bc_info.write(alias)
            first = False

        bc_info.write("\n")

File: image/build-bitcode.lua
#!/usr/bin/env lua5.3

local runas = arg[0]
function fail(msg)
    print(runas .. ": " .. msg)
    -- TODO: proper error codes.
    os.exit(1)
end

function isdir(filename)
    _, _, code = os.execute("[ -d " .. filename .. " ]")
    return code == 0
end

function basename(filename)
    return string.gsub(filename, ".*/([^/]*)", "%1")
end

local commands = {
    prt     = "prt-get",
    getbc   = "get-bc",
    dis     = "llvm-dis",
    pkginfo = "pkginfo",
    loc     = "loc",
    opt     = "opt",
    mkdir   = "mkdir",
    md5     = "md5sum",
    mv      = "mv",
    rm      = "rm"
}

local files = {
    pkgmk_conf = "/etc/pkgmk.conf",
    bc_dir     = os.getenv("BC_DIR"),
    bc_info    = nil,
    pkgs       = os.getenv("PKGS_FILE")
}

local flags = {
    c   = "",
    cxx = "",
}

local elf_magic = "\x7FELF"

if not files.bc_dir then
    fail("BC_DIR environment variable not set!")
else
    -- Bitcode info file (lines, aliases).
    files.bc_info = files.bc_dir .. "/info.txt"
end

if not os.execute(commands.mkdir .. " " .. files.bc_dir) then
    fail("Could not make BC_DIR = '" .. files.bc_dir .. "' directory!")
end

if not files.pkgs then
    fail("PKGS_FILE environment variable not set!")
end

-- TODO: check files.pkgmk exists.

-- List of pkgs in files.pkgs.
pkgs = { }
-- MD5 hash to list of installed files with that hash.
aliases = { }
-- MD5 hash to LOC for that BC.
locs = { }

-- Grab CFLAGS and CXXFLAGS explicitly. Some ports don't respect pkgmk.conf's.
for line in io.lines(files.pkgmk) do
    local s, _ = string.find(line, "%s*export CFLAGS%s*=%s*")
    if s then
        flags.c = string.sub(line, s)
    end

    s, _ = string.find(line, "%s*export CXXFLAGS%s*=%s*")
    if s then
        flags.cxx = string.sub(line, s)
    end
end

for pkg in io.lines(files.pkgs) do
    pkg = string.gsub(pkg, "%s", "")

    -- Make sure the pkg exists.
    if not os.execute(commands.prt .. " info " .. pkg) then
        fail("Bad package name: " .. pkg .. "!")
    end

    pkgs[#pkgs+1] = pkg
end

for _, pkg in pairs(pkgs) do
    -- If it's already installed, we "update", otherwise, depinst.
    local subcommand = "update"
    if not os.execute(commands.prt .. " isinst " .. pkg) then
        subcommand = "depinst"
    end

    local export = ""
    if flags.c ~= "" then export = "export CFLAGS=" .. flags.c end
    if flags.cxx ~= "" then export = export .. " export CXXFLAGS=" .. flags.cxx end
    if not os.execute(export
                      .. commands.prt
                      .. " -f -if -im -is -kw -fr --install-scripts "
                      .. subcommand .. " " .. pkg) then
       fail("Failed to build " .. pkg .. "!")
    end

    path_f = io.popen(commands.prt .. " path " .. pkg)
    local fp_path = string.gsub(path_f:read("a"), "%s", "") .. "/.footprint"
    path_f:close()
    for line in io.lines(fp_path) do
        -- Get rid of links.
        line = string.gsub(line, "%s%->%s.*$", "")
        -- Get rid of owner, size, etc..
        local installed_file = string.gsub(line, ".*%s.*%s", "")
        -- Make the file relative to root.
        installed_file = "/" .. installed_file
        print(installed_file)

        -- Don't need directories, obviously.
        if not isdir(installed_file) then
            local installed_file_f = io.open(installed_file)
            local f_magic = installed_file_f:read(4)
            installed_file_f:close()
            if f_magic == elf_magic then
                local bc = basename(installed_file) .. ".bc"
                local bc_path = files.bc_dir .. "/" .. bc

                md5_f = io.popen(commands.md5 .. " " .. installed_file)
                local md5 = string.gsub(md5_f:read("a"), "([^%s]*)%s.*", "%1")
                md5_f:close()

                if aliases[md5] ~= nil then
                    aliases[md5][#aliases[md5]+1] = bc
                else
                    aliases[md5] = { [1] = bc }
                end

                if not os.execute(commands.getbc .. " -o "
                                  .. bc_path .. " " .. installed_file) then
                    fail("Failed to get BC for '" .. installed_file .. "'!")
                end

                -- Grab .ll, for loc.
                if not os.execute(commands.dis .. " " .. bc_path
                                  .. " -o " .. bc_path .. ".ll") then
                    fail("Failed to disassemble '" .. bc "'!")
                end

                -- Get LOC from .ll.
                loc_f = io.popen(commands.loc .. " " .. bc_path .. ".ll")
                local loc = string.gsub(loc_f:read("a"), "%s", "")
                loc_f:close()

                locs[md5] = loc

                -- If BC was built with debug info, strip it.
                os.execute(commands.mv .. " " .. bc_path .. " " .. bc_path .. ".dbg")
                if not os.execute(commands.opt .. " -strip-debug " .. bc_path .. ".dbg"
                                  .. " -o " .. bc_path) then
                   fail("Failed to strip debug info for '" .. bc "'!")
                end

                -- Get rid of the .ll and .dbg
                os.execute(commands.rm .. " " .. bc_path .. ".dbg")
                os.execute(commands.rm .. " " .. bc_path .. ".ll")
            end
        end
    end
end

-- Write everything to the info file.
local info_f = io.open(files.bc_info, "w")
if not info_f then
    fail("Could not open info file!")
end

info_f:write("CFLAGS: " .. flags.c .. "\n")
info_f:write("CXXFLAGS: " .. flags.cxx .. "\n")

info_f:write("LOC\t\tBitcode\t\tAliases\n")
info_f:write("---\t\t-------\t\t-------\n")
for md5, bcs in pairs(aliases) do
    local canon = bcs[1]
    local loc = locs[md5]

    info_f:write(loc .. "\t\t" .. canon .. "\t\t")
    for i = 2, #bcs, 1 do
        if i ~= 2 then info_f:write(" ") end
        info_f:write(bcs[i])
    end

    info_f:write("\n")
end
